### PR TITLE
Multiple versions prefetch fix + gzip limit

### PR DESCRIPTION
**Pull request by Focshole**
This should fix the multiple extension issue #37 and the todo for the gzip db files, which their extraction size was unrestricted and could have caused a DoS by exhausting the storage size if a mirror would have served a gzip bomb rather than a proper db file.
I had also refactored some code because I didn't properly named some functions and structs (I did some confusion between repo and mirrors).

### CODE REVIEW

**Comment by Focshole** [comment]
As a future work, I am planning on refactoring the logic to handle the .db files, so that it becomes much similar to the packages ones.
Right now if you change mirror urls from an existing repo, it may happen that the prefetching algorithm won't sync to the changes, which is bad.
I had thought about doing another commit for that, to avoid cluttering this fix too much.

**Comment by Focshole** [comment]
Edit: I had finished the refactoring so far and made it non-breaking with older dbs, I am keeping that commit in another branch temporarily.
Maybe tomorrow I'll write the reflector thing

**Comment by anatol** [comment]
@Focshole could you please rebase your changes on top of `anatol/master` so it will avoid having unneeded merge commits in the history.

**Review by Focshole — COMMENTED** [review]
I hope it will be fine, sorry for the mess

**Comment by Focshole** [comment]
Ok, I had finished writing, checking and testing the refactoring. I had updated the prefetching logic for  `.db` files, making it like for `.pkg` files.
It is no more dependent on mirror URLs, which was an issue (if you changed URL in the config, it wouldn't be updated on the prefetching db neither after restarting it).
Also, there was an issue in which MirrorDB urls were not being deleted, so I decided to push those modifications together with the other issue fix. 

If you prefer having the two commits merged together, I'll do it

**Comment by anatol** [comment]
Could you please add more information to the commit messages? Add the context of the change as well as reference to any github issues/discussions.

The first commit in your PR is actually contains 2 logical changes "Removed multiple ext" and "fix for gzip bombs". It is always a good thing to keep distinct features/fixes in different commits. It makes it easier to review/test or revert it in the future. If you can split the commit in two parts please do, it will make the git history cleaner. Let me know if you need a help with splitting.

Otherwise the PR looks good.

File: repo_db_mirror.go [hunk 1]
```diff
@@ -28,14 +26,19 @@ func uncompressGZ(filePath string, targetFile string) error {
 		log.Printf("error: %v\n", err)
 		return err
 	}
```
Comment (Focshole) [inline repo_db_mirror.go:27]: Fix for gzip bombs

File: repo_db_mirror.go [hunk 2]
```diff
@@ -15,8 +15,6 @@ import (
 )
```
Comment (Focshole) [inline repo_db_mirror.go:14]: Beginning gzip bombs fix

File: repo_db_mirror.go [hunk 3]
```diff
@@ -28,14 +26,19 @@ func uncompressGZ(filePath string, targetFile string) error {
...
 	if err != nil {
 		log.Printf("error: %v\n", err)
 		return err
 	}
 	defer writer.Close()
-	if _, err = io.Copy(writer, reader); err != nil {
+	if _, err = io.Copy(writer, limitedReader); err != nil {
 		log.Printf("error: %v\n", err)
 		return err
 	}
```
Comment (Focshole) [inline repo_db_mirror.go:43]: Ending gzip bomb fix

File: repo_db_mirror_test.go [hunk 4]
```diff
@@ -2,6 +2,7 @@ package main
 
 import (
 	"archive/tar"
```
Comment (Focshole) [inline repo_db_mirror_test.go:4]: Beginning gzip bomb fix test

File: repo_db_mirror_test.go [hunk 5]
```diff
@@ -219,6 +219,47 @@ func TestUncompressGZ(t *testing.T) {
...
+	}
+	size := fi.Size()
+	if size >= gzipBombSize {
+		log.Fatal("It fully extracted the gzip bomb, this shouldn't happen")
+	}
+
+}
+
 func TestExtractFilenamesFromTar(t *testing.T) {
 	tmpDir := testSetupHelper(t)
```
Comment (Focshole) [inline repo_db_mirror_test.go:263]: Ending gzip bomb fix test

**Comment by Focshole** [comment]
> Could you please add more information to the commit messages? Add the context of the change as well as reference to any github issues/discussions.
I'm not sure if it is fine as I have just done. The gzip bomb had no issue opened about, there was only a TODO in the code.

For what regards the commit splitting, I hope it is fine

**Comment by anatol** [comment]
Rebased and merged to master. Thank you for your work @Focshole!

**Pull request closed without merging**